### PR TITLE
Fix invalid max key size for RC2 cipher

### LINKED ISSUE #25: RC2 max key size incompatibility
There is a compatibility issue with the maximum key size of RC2 cipher.

mcrypt says key size must be between 1 and 128 but in the polyfill it's 56, also if key size is out of bounds, it's being truncated silently.

### PR DESCRIPTION
**Pull request by umitakkaya**
Fixes #25 
- Added warning trigger when the key size has been changed by the polyfill.

### CODE REVIEW

**Comment by terrafrost** [comment]
Your fix for the immediate issue (as reported on in #25) is good but...

> * Added warning trigger when the key size has been changed by the polyfill.

The issue I have with that is that mcrypt, itself, does not issue a warning when it adjusts the key size. It might be nice if it did but it doesn't and people attempting to use this polyfill as a drop in replacement for mcrypt might, all of a sudden, start getting errors where before they weren't.

I'd say: update the PR to remove the warning. I can add a unit test that'll reproduce the immediate issue and verify the fix.

File: tests/MCryptCompatTest.php [hunk 1]
```diff
@@ -546,6 +562,7 @@ public function testMiddleKey()
         $plaintext = 'a';
 
         $td = mcrypt_module_open('rijndael-128', '', 'cbc', '');
+        $this->expectTriggerError('phpseclib_set_key(): Key size of 20 has been changed to: 24', E_USER_WARNING);
```
Comment (terrafrost) [inline tests/MCryptCompatTest.php]: This is testing the *real* mcrypt (if it's available) and using it to kinda establish base behavior. The real mcrypt doesn't output an error and, even if it did, I'm pretty sure that error wouldn't mention phpseclib lol

File: tests/MCryptCompatTest.php [hunk 2]
```diff
@@ -1030,4 +1052,30 @@ public function setExpectedException($name, $message = null, $code = null)
...
+
+    public function errorHandler($errNo, $errStr, $errFile, $errLine, $errContext)
+    {
+        if ($errStr === $this->expectedErrorMessage && $errNo === $this->expectedErrorLevel)
+        {
+            return;
+        }
+
+        call_user_func($this->previousErrorHandler, $errNo, $errStr, $errFile, $errLine, $errContext);
+    }
```
Comment (terrafrost) [inline tests/MCryptCompatTest.php]: Seems like we'd be better off re-using the existant `setExpectedException` method:

https://github.com/phpseclib/mcrypt_compat/blob/master/tests/MCryptCompatTest.php#L1014

**Comment by umitakkaya** [comment]
@terrafrost I'm not sure about other ciphers but running following code on php5.6 triggers an error, that's why I added trigger_error.

```php
$type = 'rc2';
$mode = 'ecb';
$key  = str_repeat('a', 129);
$iv   = mcrypt_create_iv(mcrypt_get_iv_size($type, $mode), MCRYPT_RAND);

$input = 'aaaaaaaaaa';

$encrypted = rtrim(
	base64_encode(
		mcrypt_encrypt($type, $key, $input, $mode, $iv)
	)
);

$decrypted = trim(
	mcrypt_decrypt(
		$type,
		$key,
		base64_decode($encrypted),
		$mode,
		mcrypt_create_iv(mcrypt_get_iv_size($type, $mode), MCRYPT_RAND)
	)
);

echo $decrypted;
```

Output:

```
$ php56 mcrypt_warning.php

PHP Warning:  mcrypt_encrypt(): Key of size 129 not supported by this algorithm. Only keys of size 1 to 128 supported in 	mcrypt_warning.php on line 12
PHP Stack trace:
PHP   1. {main}() 	mcrypt_warning.php:0
PHP   2. mcrypt_encrypt() 	mcrypt_warning.php:12

Warning: mcrypt_encrypt(): Key of size 129 not supported by this algorithm. Only keys of size 1 to 128 supported in 	mcrypt_warning.php on line 12

Call Stack:
    0.0067     124120   1. {main}() 	mcrypt_warning.php:0
    0.0071     124728   2. mcrypt_encrypt() 	mcrypt_warning.php:12

PHP Warning:  mcrypt_decrypt(): Key of size 129 not supported by this algorithm. Only keys of size 1 to 128 supported in 	mcrypt_warning.php on line 23
PHP Stack trace:
PHP   1. {main}() 	mcrypt_warning.php:0
PHP   2. mcrypt_decrypt() 	mcrypt_warning.php:23

Warning: mcrypt_decrypt(): Key of size 129 not supported by this algorithm. Only keys of size 1 to 128 supported in 	mcrypt_warning.php on line 23

Call Stack:
    0.0067     124120   1. {main}() 	mcrypt_warning.php:0
    0.0087     125184   2. mcrypt_decrypt() 	mcrypt_warning.php:23
```

**Comment by terrafrost** [comment]
> I'm not sure about other ciphers but running following code on php5.6 triggers an error, that's why I added trigger_error.

_That_ error is already handled here:

https://github.com/phpseclib/mcrypt_compat/blob/1.0.8/lib/mcrypt.php#L950

Quoting the comment from that function:

```
// PHP 5.6 made mcrypt_encrypt() a lot less tolerant of bad input but it neglected to change
// anything about mcrypt_generic(). and despite the changes insufficiently long plaintext
// is still accepted.
```
So like I said in my earlier post:

> update the PR to remove the warning. I can add a unit test that'll reproduce the immediate issue and verify the fix.

**Comment by terrafrost** [comment]
Test code (after mcrypt_compat has already been loaded):

```php
// set the params up
$key = str_repeat('z', 129);
$iv = str_repeat('z', 8);
$plaintext = 'a';

// test on mcrypt
$td = mcrypt_module_open('rc2', '', 'cbc', '');
mcrypt_generic_init($td, $key, $iv);
$mcrypt = bin2hex(mcrypt_generic($td, 'This is very important data'));

echo $mcrypt;

// test with mcrypt_compat
$td = phpseclib_mcrypt_module_open('rc2', '', 'cbc', '');
phpseclib_mcrypt_generic_init($td, $key, $iv);
$mcrypt = bin2hex(phpseclib_mcrypt_generic($td, 'This is very important data'));

echo $mcrypt;
```
Even without your RC2-specific fix the mcrypt_compat code outputs an error. The base64-encoded strings, however, won't match without your RC2-specific fix.

**Comment by terrafrost** [comment]
I took your branch, did a `git reset --hard` to the first of your two commits and then manually removed the `trigger_error` line you added. I then squashed my commit into your commit and then cherry-picked your altered commit from the master branch to the 1.0 branch and then merged the 1.0 branch into master:

https://github.com/phpseclib/mcrypt_compat/commit/f16c236f7ec8ac56ea4a3eac77df30230a8da363

Thanks!

**Pull request closed without merging**